cpp/src/arrow/util/decimal.cc

```diff
@@ -521,6 +750,23 @@ std::string Decimal128::ToString(int32_t scale) const {
   return str;
 }
 
+static inline void ShiftAndAdd(std::string_view input, uint32_t* out) {
+  const uint32_t len =
+      static_cast<uint32_t>(std::min(kInt32DecimalDigits + 1, input.size()));
+  if (len == 0) {
+    return;
+  }
+
+  uint32_t value = 0;
+  ARROW_CHECK(internal::ParseValue<UInt32Type>(input.data(), len, &value));
+
+  uint64_t tmp = *out;
+  tmp *= kUInt32PowersOfTen[len];
+  tmp += value;
+
+  *out = static_cast<uint32_t>(tmp & 0xFFFFFFFFU);
```

Comment:
So we're just losing bits if there's an overflow here? I'll note that `SimpleDecimalFromString` checks for overflows on other ops.

How about:
1. remove the `uint32_t`-output variant of this function
2. for Decimal32, call the `uint64_t` variant and reuse the existing limits check

Reply:
`SimpleDecimalFromString` has the same overflow check for both Decimal32 and Decimal64 though? 

that said, removing the `uint32_t` variant and just using the `uint64_t` variant does simplify this, so I'll do that.